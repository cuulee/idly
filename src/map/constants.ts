import { FillLayer } from 'map/layers/area';
import { LineLayer } from 'map/layers/line';
import { PointsWithLabels } from 'map/layers/pointsWithLabels';
import { PointsWithoutLabels } from 'map/layers/pointsWithoutLabel';

export const ZOOM = 16;
export const SOURCES = [
  {
    source: 'virgin',
    data: 'entities'
  },
  {
    source: 'modified',
    data: 'modifiedEntities'
  }
];

export const Layers = {
  PointsWithLabels,
  PointsWithoutLabels,
  LineLayer,
  FillLayer
};

export const SELECTABLE_LAYERS = Object.keys(Layers)
  .filter(l => Layers[l].selectable)
  .map(l => SOURCES.map(s => s.source + Layers[l].displayName))
  .reduce((prv, c) => prv.concat(c), []);

export const LAYERS = Object.keys(Layers)
  .map(l => SOURCES.map(s => s.source + Layers[l].displayName))
  .reduce((prv, c) => prv.concat(c), []);
