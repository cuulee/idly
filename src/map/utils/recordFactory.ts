import { Record } from 'immutable';

type Updater<T> = (value: T) => T;
/**
 * This generates a record constructor which is
 * typed strong.:)
 */

export interface IRecord<T> {
  get<K extends keyof T>(key: K): T[K];
  set<K extends keyof T>(key: K, value: T[K]): IRecord<T>;
  update<K extends keyof T>(key: K, updater: Updater<T[K]>): IRecord<T>;

  // Deep operations still need to be manually defined.
  getIn(keyPath: Array<number | string>, notSet?: any): any | undefined;
  setIn(keyPath: string[], value: any): IRecord<T>;
  deleteIn(keyPath: string[]): IRecord<T>;

  withMutations(mutator: (s: IRecord<T>) => any): IRecord<T>;

  // Merge is made easy using typescript's new mapped types!!!
  merge(partial: Partial<T>): IRecord<T>;
  mergeDeep(partial: Partial<T>): IRecord<T>;

  equals(other: IRecord<T>): boolean;
  toJS(): any;
}
export function recordFactory<T>(defaultState: T, recordName?: string) {
  const RecordClass = Record(defaultState, recordName);
  interface IConstructor<TInput> {
    (input: TInput): IRecord<T>;
    new (input: TInput): IRecord<T>;
  }
  return (RecordClass as any) as IConstructor<T>;
}
