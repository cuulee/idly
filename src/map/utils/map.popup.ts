import * as _ from 'lodash';
import { LAYERS } from 'map/constants';
import { popup } from 'map/mapboxglSetup';
import * as turf from 'turf';
const SIZE = 10;
export const dirtyPopup = map =>
  LAYERS.map(s => {
    map.on(
      'mouseenter',
      s,
      _.debounce(e => {
        // Change the cursor style as a UI indicator.
        map.getCanvas().style.cursor = 'pointer';
        const bbox = [
          [e.point.x - SIZE, e.point.y - SIZE],
          [e.point.x + SIZE, e.point.y + SIZE]
        ];
        if (map.queryRenderedFeatures(bbox, { layers: [s] }).length > 0) {
          map.getCanvas().style.cursor = 'pointer';
        }
        if (!Array.isArray(e.features) || !e.features[0]) return;
        const html = e.features.map(feat => ({
          ...feat,
          id: feat.properties.id,
          tags: JSON.parse(feat.properties.tags),
          node_properties: JSON.parse(
            feat.properties.node_properties || feat.properties.way_properties
          ),
          icon: feat.properties.icon
        }));
        popup
          .setLngLat(turf.centroid(e.features[0]).geometry.coordinates)
          .setHTML(
            `<span style="width:200px"><pre>${JSON.stringify(
              {
                id: e.features.map(f => f.properties.id),
                tags: e.features.map(f => JSON.parse(f.properties.tags))
              },
              null,
              2
            )}</pre>
          </span>`
          )
          .addTo(map);
      }),
      800
    );
    map.on(
      'mouseleave',
      s,
      _.debounce(() => {
        map.getCanvas().style.cursor = '';
        popup.remove();
      }, 100)
    );
  });
