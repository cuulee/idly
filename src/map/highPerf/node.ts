import { Feature, Point } from 'geojson';
import { Set } from 'immutable';
import { Geometry } from 'osm/entities/constants';
import { Tags } from 'osm/entities/helpers/tags';
import { Node } from 'osm/entities/node';
import { ParentWays } from 'osm/parsers/parsers';
import { presetsMatcher, presetsMatcherCached } from 'osm/presets/presets';
import * as R from 'ramda';
import { weakCache, weakCache2 } from 'utils/weakCache';
type NodeFeat = Feature<Point>;
type NodeGeometry = Geometry.POINT | Geometry.VERTEX | Geometry.VERTEX_SHARED;

interface INodeMarkup {
  icon: string;
  name?: string;
  geometry: NodeGeometry;
}
export const DEFAULT_NODE_ICON = 'circle';

export function nodeCombiner(node: Node, parentWays: ParentWays) {
  return {
    ...nodeToPoint(node),
    properties: {
      ...applyNodeMarkup(
        // presetsMatcher,
        getNodeGeometry(node.id, parentWays),
        node.tags
      ),
      id: node.id
    }
  };
}

export const nodeToPoint = (node: Node): Feature<Point> => {
  return {
    id: node.id,
    type: 'Feature',
    geometry: {
      type: 'Point',
      coordinates: [node.loc.lon, node.loc.lat]
    },
    properties: {}
  };
};

export const applyNodeMarkup = (geometry: NodeGeometry, tags: Tags) => {
  const match = presetsMatcherCached(geometry)(tags);
  return {
    icon: (match && match.icon) || DEFAULT_NODE_ICON,
    name: tags.get('name'),
    geometry
  };
};

/**
 * @NOTE
 *  The way osm works is that it will give you everything inside the bbox
 *   but only a secondary level outside the bbox. So if the a node is
 *   currently a vertex, it could become a vertex_shared in future.
 *   thats why parentWays is updated on every network request.
 */
export function getNodeGeometry(id, parentWays: ParentWays) {
  if (parentWays.has(id))
    return parentWays.get(id).size > 1
      ? Geometry.VERTEX_SHARED
      : Geometry.VERTEX;
  return Geometry.POINT;
}
