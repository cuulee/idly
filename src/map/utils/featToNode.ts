import { Node, nodeFactory } from 'osm/entities/node';
import { genLngLat } from 'osm/geo_utils/lng_lat';
import { propertiesGen } from 'osm/others/properties';
import { tagsFactory } from 'osm/others/tags';

import { NodeFeature } from 'map/utils/nodeToFeat';

export function featToNode(feat: NodeFeature): Node {
  return nodeFactory({
    id: feat.properties.id,
    tags: tagsFactory(JSON.parse(feat.properties.tags)),
    loc: genLngLat([
      feat.geometry.coordinates[0],
      feat.geometry.coordinates[1]
    ]),
    properties: propertiesGen(JSON.parse(feat.properties.node_properties))
  });
}
