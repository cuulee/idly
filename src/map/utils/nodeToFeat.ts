import { Point } from 'geojson';
import * as turf from 'turf';

import { Node } from 'osm/entities/node';
import { ITags } from 'osm/entities/helpers/tags';
import { presetsMatch } from 'osm/presets/presets';
import { Feature } from 'typings/geojson';
import { weakCache } from 'utils/weakCache';

interface INodeProperties {
  node_properties: string;
  tags: string;
  id: string;
  icon: string;
}
/**
 * @TOFIX this whole Nodeproperties fuck up
 *  the typings dont support properties
 */
export type NodeFeature = Feature<Point, INodeProperties>;

function _nodeToFeat(n: any): NodeFeature {
  if (n instanceof Node) {
    const match = presetsMatch(n);
    const properties: INodeProperties = {
      node_properties: JSON.stringify(n.properties),
      tags: JSON.stringify(n.tags),
      id: n.id,
      icon: (match && match.icon) || 'circle'
    };
    const feat = turf.point([n.loc.lon, n.loc.lat], properties) as Feature<
      Point,
      INodeProperties
    >;
    feat.id = n.id;
    return feat;
  }
}

export const nodeToFeat = weakCache(_nodeToFeat);
