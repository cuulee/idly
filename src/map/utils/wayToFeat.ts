import { LineString, Polygon } from 'geojson';
import * as turf from 'turf';

import { Feature } from 'typings/geojson';

import { getGeometry } from 'osm/entities/helpers/misc';
import { Node } from 'osm/entities/node';
import { initAreaKeys, initPresets, presetsMatch } from 'osm/presets/presets';

import { Geometries } from 'osm/entities/constants';
import { Way } from 'osm/entities/way';
import { Graph } from 'osm/history/graph';
import { weakCache, weakCache2 } from 'utils/weakCache';

interface INodeProperties {
  node_properties: string;
  tags: string;
  id: string;
  name?: string;
  geometry: string;
}
/**
 * @TOFIX this whole Nodeproperties fuck up
 *  the typings dont support properties
 */
export type WayFeature = Feature<LineString | Polygon, INodeProperties>;
const { collection } = initPresets();
const areaKeys = initAreaKeys(collection);
/**
 *
 * @REVISIT this func also needs to handle modifiedGraph
 *  for now i removed it so i can test caching and anyway
 *  not yet implemented modified ways
 */
function _wayToFeat(w: Way, graph: Graph): WayFeature {
  if (w instanceof Way) {
    const match = presetsMatch(w, areaKeys);
    const properties: INodeProperties = {
      node_properties: JSON.stringify(w.properties),
      tags: JSON.stringify(w.tags),
      id: w.id,
      geometry: w.geometry
    };
    const nodes = w.nodes.map(id => {
      return graph.node.get(id); // || graphMod.node.get(id);
    });
    let feat;
    if (w.geometry === Geometries.LINE) {
      feat = turf.lineString(
        nodes.map(n => [n.loc.lon, n.loc.lat]).toArray(),
        properties
      ) as Feature<LineString, INodeProperties>;
      feat.id = w.id;
    } else if (w.geometry === Geometries.AREA) {
      const geos = nodes.map(n => [n.loc.lon, n.loc.lat]).toArray();
      geos.push(geos[0]);
      feat = turf.polygon([geos], properties) as Feature<
        Polygon,
        INodeProperties
      >;
    }
    feat.id = w.id;
    return feat;
  }
}

export const wayToFeat = weakCache2(_wayToFeat);
