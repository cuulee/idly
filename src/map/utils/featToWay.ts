import { List } from 'immutable';

import { propertiesGen } from 'osm/entities/helpers/properties';
import { tagsFactory } from 'osm/entities/helpers/tags';
import { Node } from 'osm/entities/node';
import { genLngLat } from 'osm/geo_utils/lng_lat';

import { WayFeature } from 'map/utils/wayToFeat';
import { wayFactory } from 'osm/entities/way';

import { Graph } from 'osm/history/graph';

export function featToWay(feat: WayFeature, graph: Graph) {
  const nodesId = JSON.parse(feat.properties.nodes);
  const nodes = nodesId.map((n, i) => {
    const node: Node = graph.node.get(n);
    return node.set('loc', genLngLat(feat.geometry.coordinates[i]));
  });

  return [
    wayFactory({
      id: feat.properties.id,
      tags: tagsFactory(JSON.parse(feat.properties.tags)),
      properties: propertiesGen(JSON.parse(feat.properties.way_properties)),
      geometry: feat.properties.geometry,
      nodes: List(nodesId)
    }),
    ...nodes
  ];
}
