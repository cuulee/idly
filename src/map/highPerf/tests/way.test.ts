import { List } from 'immutable';

import { tagsFactory } from 'osm/entities/helpers/tags';
import { nodeFactory } from 'osm/entities/node';
import { wayFactory } from 'osm/entities/way';
import { graphFactory } from 'osm/history/graph';

import {
  getCoordsFromGraph,
  wayCombiner,
  wayToLineString
} from 'map/highPerf/way';
import { Geometry } from 'osm/entities/constants';
import { genLngLat } from 'osm/geo_utils/lng_lat';
import { getWayGeometry } from 'osm/parsers/parsers';
import { wayToFeat } from 'map/utils/wayToFeat';

const way = wayFactory({
  id: 'w1',
  nodes: List(['n1']),
  tags: tagsFactory({ highway: 'residential' })
});

const node = nodeFactory({ id: 'n1' });

const graph = graphFactory([node, way]);

describe('way.test', () => {
  describe('wayToLineString', () => {
    it('should throw error when no geometry provided', () => {
      expect(() => wayToLineString(null, [[1, 2]])).toThrow();
      expect(() => wayToLineString(Geometry.POINT, [[1, 2]])).toThrow();
    });
    it('should not work when correct coords provided with wrong geometry', () => {
      expect(() =>
        wayToLineString(Geometry.POINT, [[1, 2], [2, 3], [3, 0], [1, 2]])
      ).toThrow();
    });
    it('should throw error when wrong coords provided', () => {
      expect(() => wayToLineString(Geometry.AREA, [[1, 2], [2, 1]])).toThrow();
    });
    it('should work when correct coords provided', () => {
      expect(() =>
        wayToLineString(Geometry.AREA, [[1, 2], [2, 3], [3, 0], [1, 2]])
      ).not.toThrow();
    });
    it('should behave...', () => {
      expect(
        wayToLineString(Geometry.AREA, [[1, 2], [2, 3], [3, 0], [1, 2]])
      ).toMatchSnapshot();
    });
  });

  describe('getNodesFromGraph', () => {
    it('should get nodes correctly', () => {
      const n1 = nodeFactory({ id: 'n1', loc: genLngLat([1, 2]) });
      const g = graphFactory([n1]);
      expect(getCoordsFromGraph(g.node, List(['n1']))).toMatchSnapshot();
    });
    it('should throw error if nodes not found', () => {
      const n1 = nodeFactory({ id: 'n1', loc: genLngLat([1, 2]) });
      const g = graphFactory([n1]);
      expect(() => getCoordsFromGraph(g.node, List(['n2']))).toThrow();
    });

    it('should work when multiple nodes are there ingraph', () => {
      const n1 = nodeFactory({ id: 'n1', loc: genLngLat([1, 2]) });
      const n2 = nodeFactory({ id: 'n2', loc: genLngLat([3, 4]) });
      const g = graphFactory([n1, n2]);
      expect(getCoordsFromGraph(g.node, List(['n2']))).toMatchSnapshot();
      expect(getCoordsFromGraph(g.node, List(['n1', 'n2']))).toMatchSnapshot();
    });
  });

  describe('wayCombiner', () => {
    const n1 = nodeFactory({ id: 'n1', loc: genLngLat([1, 2]) });
    const n2 = nodeFactory({ id: 'n2', loc: genLngLat([3, 4]) });
    const n3 = nodeFactory({ id: 'n3', loc: genLngLat([1, 4]) });

    const w1 = wayFactory({
      id: 'w1',
      nodes: List(['n1', 'n3']),
      tags: tagsFactory({ highway: 'residential' })
    });

    const g = graphFactory([n1, n2, n3]);
    it('should behave...', () => {
      const result = wayCombiner(way, graph);
      expect(result).toMatchSnapshot();
      const result2 = wayCombiner(w1, g);
      expect(result2).toMatchSnapshot();
    });

    it('should take name', () => {
      const w2 = wayFactory({
        id: 'w1',
        nodes: List(['n1']),
        tags: tagsFactory({ highway: 'residential', name: 'great highway' })
      });
      const g2 = graphFactory([node, w2]);
      const result = wayCombiner(w2, g2);
      expect(result.properties.name).toEqual('great highway');
    });
  });
});
