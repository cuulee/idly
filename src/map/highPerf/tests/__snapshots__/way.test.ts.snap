// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`way.test getNodesFromGraph should get nodes correctly 1`] = `
Array [
  Array [
    1,
    2,
  ],
]
`;

exports[`way.test getNodesFromGraph should work when multiple nodes are there ingraph 1`] = `
Array [
  Array [
    3,
    4,
  ],
]
`;

exports[`way.test getNodesFromGraph should work when multiple nodes are there ingraph 2`] = `
Array [
  Array [
    1,
    2,
  ],
  Array [
    3,
    4,
  ],
]
`;

exports[`way.test wayCombiner should behave... 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        NaN,
        NaN,
      ],
    ],
    "type": "LineString",
  },
  "properties": Object {
    "geometry": "line",
    "icon": "highway-residential",
    "id": "w1",
    "name": undefined,
    "tagsClass": "tag-highway",
    "tagsClassType": "tag-highway-residential",
  },
  "type": "Feature",
}
`;

exports[`way.test wayCombiner should behave... 2`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        1,
        2,
      ],
      Array [
        1,
        4,
      ],
    ],
    "type": "LineString",
  },
  "properties": Object {
    "geometry": "line",
    "icon": "highway-residential",
    "id": "w1",
    "name": undefined,
    "tagsClass": "tag-highway",
    "tagsClassType": "tag-highway-residential",
  },
  "type": "Feature",
}
`;

exports[`way.test wayToLineString should behave... 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        Array [
          1,
          2,
        ],
        Array [
          2,
          3,
        ],
        Array [
          3,
          0,
        ],
        Array [
          1,
          2,
        ],
      ],
    ],
    "type": "Polygon",
  },
  "properties": Object {},
  "type": "Feature",
}
`;
